src/groupeddataframe/complextransforms.jl

```diff
@@ -273,11 +273,11 @@ function _combine_rows_with_first!((firstrow,)::Ref{Any},
     tasks = Vector{Task}(undef, length(partitions))
     for (tid, idx) in enumerate(partitions)
         tasks[tid] =
-            @spawn _combine_rows_with_first_task!(tid, first(idx), last(idx), first(idx),
-                                                  outcols, outcolsref,
-                                                  type_widened, widen_type_lock,
-                                                  f, gd, starts, ends, incols, colnames,
-                                                  firstcoltype(firstmulticol))
+            @spawn_or_async _combine_rows_with_first_task!(tid, first(idx), last(idx), first(idx),
```

Comment:
a lot of code that isn't thread-safe is also not safe for `@async` concurrency. Silly example:
```julia
julia> v = []
Any[]

julia> @sync for i = 1:100
           @async begin
               sleep(rand())
               push!(v, i)
           end
       end

julia> issorted(v)
false
```
Perhaps folks will push an ID to a list or something inside a `combine`, and expect `singlethreaded()` will give no concurrency so it works?

My actual concern is loading data with Downloads.jl or HTTP.jl; both have a lot of issues with concurrency right now, and even `@async` is enough to quickly lock up release versions of HTTP.jl. So if you do something like "for each group, download a file associated to that group and compute some statistic of it" (a thing I actually do) then this will have issues.

So I think it would be great if there was a way to disallow concurrent execution of a function altogether, from green threads or threads.

Reply:
@jrevels pointed out we often don't really want to refer to threads but rather tasks, which kind of generalizes my point above. Maybe it should be `singletask` instead of `singlethreaded`? (or `NotPure` as suggested in the issue)

Reply:
This is good point - I think `@async` should be also disabled.

Reply:
Ah, too bad. I hoped we could support this without adding branches all over the place which make the code hard to read and add special code paths that need a specific test for each case. This applies in particular to splitapplycombine.jl:669. The only way to avoid adding branches I can see is having the `@run_or_async` macro run the code immediately, and return a pseudo-`Task` object that supports `with` and `fetch` but makes them no-ops. But that's quite complex too.

Reply:
Actually we can simply run the passed expression and return a `Task` that returns its value. It's a bit of a waste but not worse than spawning multiple tasks when the number of threads is 1. The cost is negligible anyway, which I why we always spawn tasks in these functions.